metaflow/plugins/aws/batch/batch.py

```diff
@@ -53,15 +54,26 @@ def _search_jobs(self, flow_name, run_id, user):
         return jobs
 
     def _job_name(self, user, flow_name, run_id, step_name, task_id, retry_count):
-        return '{user}-{flow_name}-{run_id}-{step_name}-{task_id}-{retry_count}'.format(
-            user=user,
+        job_name = '{user}-{flow_name}-{run_id}-{step_name}-{task_id}-{retry_count}'.format(
+            user=user.replace(" ", ""),
             flow_name=flow_name,
             run_id=run_id,
             step_name=step_name,
             task_id=task_id,
             retry_count=retry_count,
         )
 
+        # AWS Batch job name validation according to
+        # https://docs.aws.amazon.com/batch/latest/APIReference/API_SubmitJob.html#API_SubmitJob_RequestSyntax
+        if not job_name[0].isalnum():
+            raise BatchException('The Batch Jobs first character has to be alphanumeric')
```

Comment:
Users of metaflow aren't aware of the notion of a `Batch Job Name`. Throwing this error message will most likely confuse the users since they wouldn't necessarily know how to fix it. 

One thought is to have a check in the [Batch decorator class](https://github.com/Netflix/metaflow/blob/master/metaflow/plugins/aws/batch/batch_decorator.py#L111) disallowing usernames which don't begin with an alphanumeric character, have characters which are not dashes, underscores or alphanumeric (along with a helpful error message telling them how to set a new username). The job name that metaflow creates relies on user specified username, flowname and stepname - since flowname and stepname are deduced from the flow itself and are [valid lexical identifiers](https://docs.python.org/3/reference/lexical_analysis.html#identifiers), the check on username should be good enough. 
We can maintain the check on the length of the job name, with a helpful error message asking users to reduce the size of their flow name or the step name.